File: src/services/workspace.ts
/**
 * Workspace IPC 薄封装
 *
 * 所有 Tauri invoke 调用集中在这里，UI 与 store 层只调用这些函数。
 */
import { invoke } from "@tauri-apps/api/core";
import type { WorkspaceFile, RecentWorkspaces } from "@/types/workspace";

/** 从磁盘读取 workspace 文件 */
export async function loadWorkspaceFile(path: string): Promise<WorkspaceFile> {
  return invoke<WorkspaceFile>("load_workspace_file", { path });
}

/** 原子保存 workspace 到磁盘 */
export async function saveWorkspaceFile(
  path: string,
  workspace: WorkspaceFile,
): Promise<void> {
  return invoke<void>("save_workspace_file", { path, workspace });
}

/** 新建一个空 workspace 文件（返回写入的内容） */
export async function createWorkspaceFile(
  path: string,
  name: string,
): Promise<WorkspaceFile> {
  return invoke<WorkspaceFile>("create_workspace_file", { path, name });
}

/** 最近打开过的 workspace 列表 + lastOpenedPath */
export async function listRecentWorkspaces(): Promise<RecentWorkspaces> {
  return invoke<RecentWorkspaces>("list_recent_workspaces");
}

/** 从 recent 列表里移除一项（文件被删 / 用户主动清理） */
export async function removeRecentWorkspace(path: string): Promise<void> {
  return invoke<void>("remove_recent_workspace", { path });
}

/**
 * 读取启动时缓存的 workspace 路径（来自命令行参数或 Finder 双击）
 *
 * 读完会被后端清空 —— 只生效一次。
 */
export async function getStartupWorkspacePath(): Promise<string | null> {
  return invoke<string | null>("get_startup_workspace_path");
}

/** 根据窗口 label 查询该窗口绑定的 workspace 路径 */
export async function getWindowWorkspacePath(
  label: string,
): Promise<string | null> {
  return invoke<string | null>("get_window_workspace_path", { label });
}

/** 在窗口中打开一个 workspace；已打开则 focus，否则新建窗口 */
export async function openWorkspaceInWindow(path: string): Promise<void> {
  return invoke<void>("open_workspace_in_window", { path });
}

/**
 * 把当前 workspace 的 project 路径列表推送给后端。
 *
 * fs.rs 的路径沙箱读的是 WorkspaceManager 的内存注册表 —— 没同步过的话
 * read_file / list_dir / watch_dir 都会被拒。任何会改变 projects 数组
 * 的操作（load workspace / add / remove）之后都需要调一次。
 */
export async function syncWorkspaceProjectPaths(
  paths: string[],
): Promise<void> {
  return invoke<void>("sync_workspace_project_paths", { paths });
}

/** 父目录扫描结果 —— 对照后端 ScanResult 枚举 */
export type ScanGitReposResult =
  | { kind: "single"; path: string; name: string }
  | {
      kind: "group";
      parentPath: string;
      parentName: string;
      members: Array<{ name: string; path: string }>;
    }
  | { kind: "empty" };

/**
 * 智能扫描一个用户选中的目录：
 * - 本身是 git 仓库 → `single`
 * - 本身不是 git 但一层子目录里有若干 git 仓库 → `group`
 * - 两者都不是 → `empty`
 *
 * 不走路径沙箱（因为目标还没被加进 workspace 的 allow-list）。
 */
export async function scanGitRepos(
  path: string,
): Promise<ScanGitReposResult> {
  return invoke<ScanGitReposResult>("scan_git_repos", { path });
}

File: src-tauri/src/commands/workspace.rs
//! Workspace 相关 Tauri 命令

use crate::services::workspace_manager::{
    canonicalize_or_raw, StartupPathCache, WorkspaceManager,
};
use crate::types::workspace::RecentWorkspaces;
use base64::Engine;
use serde_json::Value;
use tauri::{AppHandle, Manager, State};

/// 读取任意路径的 workspace 文件
#[tauri::command]
pub async fn load_workspace_file(
    state: State<'_, WorkspaceManager>,
    path: String,
) -> Result<Value, String> {
    state.load_file(&path)
}

/// 保存 workspace 文件（原子写），写完后更新 recent 索引
#[tauri::command]
pub async fn save_workspace_file(
    app: AppHandle,
    state: State<'_, WorkspaceManager>,
    path: String,
    workspace: Value,
) -> Result<(), String> {
    state.save_file(&path, &workspace)?;
    let name = workspace
        .get("name")
        .and_then(|v| v.as_str())
        .unwrap_or("Workspace")
        .to_string();
    if let Err(e) = state.push_recent(&app, &path, &name) {
        log::warn!("save_workspace_file: push_recent 失败: {}", e);
    }
    Ok(())
}

/// 创建一个空 workspace 文件并落盘
#[tauri::command]
pub async fn create_workspace_file(
    app: AppHandle,
    state: State<'_, WorkspaceManager>,
    path: String,
    name: String,
) -> Result<Value, String> {
    let workspace = state.create_file(&path, &name)?;
    if let Err(e) = state.push_recent(&app, &path, &name) {
        log::warn!("create_workspace_file: push_recent 失败: {}", e);
    }
    Ok(workspace)
}

/// 列出最近打开的 workspace
#[tauri::command]
pub async fn list_recent_workspaces(
    app: AppHandle,
    state: State<'_, WorkspaceManager>,
) -> Result<RecentWorkspaces, String> {
    Ok(state.load_recent(&app))
}

/// 从 recent 索引中移除一项
#[tauri::command]
pub async fn remove_recent_workspace(
    app: AppHandle,
    state: State<'_, WorkspaceManager>,
    path: String,
) -> Result<(), String> {
    state.remove_recent(&app, &path)
}

/// 读取启动参数指定的 workspace 路径（读一次即清空）
#[tauri::command]
pub async fn get_startup_workspace_path(
    cache: State<'_, StartupPathCache>,
) -> Result<Option<String>, String> {
    Ok(cache.take())
}

/// 查询某个窗口 label 对应的 workspace 路径
#[tauri::command]
pub async fn get_window_workspace_path(
    state: State<'_, WorkspaceManager>,
    label: String,
) -> Result<Option<String>, String> {
    Ok(state.path_for_label(&label))
}

/// 前端同步当前所有已注册项目根路径，用于 fs 命令的沙箱校验
#[tauri::command]
pub async fn sync_workspace_project_paths(
    state: State<'_, WorkspaceManager>,
    paths: Vec<String>,
) -> Result<(), String> {
    state.sync_project_paths(paths);
    Ok(())
}

/// 打开 workspace 到窗口：如果目标 workspace 已有窗口则 focus，
/// 否则新建 WebviewWindow，URL 带 `?ws={urlsafe_b64(path)}`。
#[tauri::command]
pub async fn open_workspace_in_window(
    app: AppHandle,
    state: State<'_, WorkspaceManager>,
    path: String,
) -> Result<(), String> {
    use tauri::{WebviewUrl, WebviewWindowBuilder};

    let canonical = canonicalize_or_raw(&path);
    let label = state.compute_label(&canonical);

    // 复用已存在的窗口
    if let Some(existing) = app.get_webview_window(&label) {
        existing.set_focus().map_err(|e| e.to_string())?;
        state.bind_window(&label, &canonical);
        return Ok(());
    }

    // 读 workspace name 用于标题
    let name = state
        .load_file(&canonical)
        .ok()
        .and_then(|v| v.get("name").and_then(|n| n.as_str()).map(|s| s.to_string()))
        .unwrap_or_else(|| "Star Gazer".to_string());

    // URL-safe base64 供前端解析
    let b64 = base64::engine::general_purpose::URL_SAFE_NO_PAD.encode(canonical.as_bytes());
    let url = format!("index.html?ws={}", b64);

    WebviewWindowBuilder::new(&app, &label, WebviewUrl::App(url.into()))
        .title(format!("Star Gazer · {}", name))
        .inner_size(1440.0, 900.0)
        .min_inner_size(1024.0, 640.0)
        .title_bar_style(tauri::TitleBarStyle::Overlay)
        .hidden_title(true)
        .resizable(true)
        .center()
        .build()
        .map_err(|e| e.to_string())?;

    state.bind_window(&label, &canonical);
    state.push_recent(&app, &canonical, &name)?;

    Ok(())
}
